### PR TITLE
MRG: cHPI position estimation

### PR DESCRIPTION
**Pull request by Eric89GXL**
Let's see if the CIs are happy. Private for now so I have a release cycle (next one) to work with it a bit more. But in the meantime acts as a nice validation of our raw simulator.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
FYI this also gets us a lot closer to movement compensation in our Maxwell filtering @wronk (#2277)

**Comment by Eric89GXL** [comment]
cc @staulu in case you want to have a look at how head position estimation is currently done in Python. I'm not 100% sure about the meaning of all the parameters MaxFilter spits out during computation, e.g.:
```
Hpi fit OK, movements [mm/s] =  0.4 / 1.9 / 3.2 / 1.6 / 1.7 /
  -63.6   48.9   87.3 / -63.7   49.6   86.8 / g = 1.000 err =  0.8 mm / 0.997 -0.078  0.028
   53.7   83.1   70.5 /  52.9   83.2   70.1 / g = 1.000 err =  0.9 mm / 0.081  0.986 -0.145
   -9.9  -25.4  138.5 / -10.5  -27.2  138.6 / g = 0.999 err =  1.9 mm /-0.017  0.146  0.989
   23.6   36.0  128.3 /  24.4   37.9  128.6 / g = 1.000 err =  2.1 mm /   7.5  -19.8   74.7
   62.7   -8.4  117.1 /  63.3   -9.4  117.4 / g = 1.000 err =  1.3 mm /
#t = 9.000, #e = 0.14 cm, #g = 1.000, #v = 0.00 cm/s, #r = 0.00 rad/s, #d = 0.00 cm
```
I have tried to recreate some of them here, but we can discuss it later. For now the function is private (and these outputs are for 'debug' mode only currently) so it can wait, unless you feel like specifying. Specifically, the ones I'm not sure about are:

1. `g` and `err` for each coil
2. `#g` for the final line
3. Whether `#d` is computed relative to the original position, or the last position.

File: mne/io/constants.py [hunk 1]
```diff
@@ -136,10 +136,11 @@ def __setattr__(self, attr, val):
 FIFF.FIFF_DESCRIPTION    = FIFF.FIFF_COMMENT # (Textual) Description of an object
 FIFF.FIFF_DIG_STRING     = 234          # String of digitized points
 FIFF.FIFF_LINE_FREQ      = 235    # Line frequency
-FIFF.FIFF_CUSTOM_REF     = 236    # Whether a custom reference was applied to the data
+FIFF.FIFF_CUSTOM_REF     = 236    # Whether a custom reference was applied to the data (whoops, overlaps...)
```
Comment (agramfort) [inline mne/io/constants.py]: ?

File: mne/chpi.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+    """Helper objective function"""
+    rot = _quat_to_rot(x[:3]).T
+    d = np.dot(est_pos_dev, rot) + x[3:] - hpi_head_rrs
+    return np.sum(d * d)
+
+
+# def _chpi_constraint(x):
+#     """Constrain quaternions to be unit length"""
+#     pos = 1. - np.sum(x[:3] * x[:3])
+#     return pos
```
Comment (agramfort) [inline mne/chpi.py]: delete?

File: mne/chpi.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+    # hpi_res_dig = hpi_result['dig_points']
+    # hpi_res_rrs = apply_trans(hpi_result['coord_trans'],
+    #                           np.array([h['r'] for h in hpi_res_dig]))
+    # assert hpi_result['coord_trans']['to'] == FIFF.FIFFV_COORD_HEAD
+    # assert hpi_result['coord_trans']['from'] == FIFF.FIFFV_COORD_DEVICE
+    # assert all(h['kind'] == FIFF.FIFFV_POINT_HPI for h in hpi_res_dig)
+    # print(1000 * np.sqrt(np.sum((hpi_res_rrs - hpi_rrs) ** 2, axis=1)))
+    # print(hpi_rrs)
+    # print(hpi_res_rrs)
+    # np.testing.assert_allclose(hpi_res_rrs, hpi_rrs)
```
Comment (agramfort) [inline mne/chpi.py]: delete?

File: mne/chpi.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+        be automatically determined. By default, a sampling interval of
+        1 second is used if processing a raw data. If processing a
+        Maxfilter log file, this must be None because the log file
+        itself will determine the sampling interval.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    translation : array
```
Comment (agramfort) [inline mne/chpi.py]: I would put the shape in the first line

translation : array, shape (n_times, 3)

File: mne/chpi.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+        A 2-dimensional array of head position vectors (n_time x 3).
+    rotation : array
+        A 3-dimensional array of rotation matrices (n_time x 3 x 3).
+    t : array
```
Comment (agramfort) [inline mne/chpi.py]: t : array, shape (n_times,)

len 1 tuple

**Comment by agramfort** [comment]
travis is not happy

how good is the coverage?

let me know when it's ready for merge.

thanks @Eric89GXL

File: mne/io/constants.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/io/constants.py]: We need to be more careful choosing new FIFF constants. The custom ref one
we chose was already occupied by the hpi freq (both are now 236)

Comment (agramfort) [inline mne/io/constants.py]: crap... we should change this...

**Comment by Eric89GXL** [comment]
@agramfort comments addressed and CIs are happy (one Travis build failed on a FieldTrip buffer problem, I restarted it). Coverage of `mne/chpi.py` is 99% (2 missed lines). Ready for merge from my end.

We can change the constant number in a subsequent PR. It'll be easier to review that way.

**Comment by staulu** [comment]
Regarding the MaxFilter parameters, g is the goodness-of-fit of the magnetic dipole forward model as compared to the extracted HPI signal vector (one g-value per HPI coil). As for the other parameters you asked for, I think I need to consult Jukka Nenonen to know for sure. I only have assumptions about them.

**Comment by Eric89GXL** [comment]
@staulu thanks, that makes sense.

File: mne/__init__.py [hunk 6]
```diff
@@ -89,6 +90,7 @@
 from . import epochs
 from . import externals
 from . import io
+io.get_chpi_positions = get_chpi_positions  # for backward compat
```
Comment (agramfort) [inline mne/__init__.py]: say in which release to remove

**Comment by agramfort** [comment]
besides LGTM

thanks @Eric89GXL

File: mne/io/constants.py (hunk 1 above)
Comment (jnenonen) [inline mne/io/constants.py]: Hi guys,
We should do some concerted work here with Elekta to keep the definitions in line with the FIFF format documentation. The current documentation (as well as neuromag libraries and source codes) can be found in the fiff_support package at http://www.aston.ac.uk/lhs/research/centres-facilities/brain-centre/facilities-clinical-services/meg-studies/downloads/. We are already defining some new extensions...

Comment (agramfort) [inline mne/io/constants.py]: oups... let's follow up by emails

**Comment by Eric89GXL** [comment]
@agramfort done and CIs are happy. Ready for merge from my end.

Also, in the last commit I improved our docstring checker to auto-ignore any function decorated with `@deprecated` so that people no longer have to add the names to the ignore list.

**Comment by agramfort** [comment]
amazing.

thanks @Eric89GXL !

**Pull request merged**